### PR TITLE
Fix access to whole sub-structs

### PR DESCRIPTION
**Pull request by kamilrakoczy**
This PR fixes access to whole sub-structs and allows to declare struct custom type field with range.

### CODE REVIEW

File: frontends/ast/simplify.cc [hunk 1]
```diff
@@ -300,13 +300,40 @@ static int size_packed_struct(AstNode *snode, int base_offset)
 	bool is_union = (snode->type == AST_UNION);
 	int offset = 0;
 	int packed_width = -1;
+	// embeded struct or union with range?
+	std::vector<AstNode *> ranges;
+	for(auto it = snode->children.begin(); it != snode->children.end(); ) {
```
Comment (kbieganski) [inline frontends/ast/simplify.cc]: Consider something like this:
```cpp
auto it = std::remove_if(snode->children.begin(), snode->children.end(),
                         [](AstNode* node) { return node->type == AST_RANGE });
std::vector<AstNode*> ranges(it, snode->children.end());
snode->children.erase(it, snode->children.end());
```
It's a bit clearer to me. Not sure if they like this sort of thing upstream.

File: frontends/ast/simplify.cc [hunk 2]
```diff
@@ -1224,6 +1249,15 @@ bool AstNode::simplify(bool const_fold, bool at_zero, bool in_lvalue, int stage,
 
 	case AST_PARAMETER:
 	case AST_LOCALPARAM:
+		if (children[0]->type == AST_IDENTIFIER) {
```
Comment (kbieganski) [inline frontends/ast/simplify.cc]: Why not combine the two outer `if`s?

**Review by kbieganski — CHANGES_REQUESTED** [review]

File: frontends/ast/simplify.cc (hunk 2 above)
Comment (kamilrakoczy) [reply frontends/ast/simplify.cc]: Fixed, I moved them to single ``if``.

File: frontends/ast/simplify.cc (hunk 1 above)
Comment (kamilrakoczy) [reply frontends/ast/simplify.cc]: True, it is also more clearer to me. I've also changed error message if there is more then single range.

**Review by kbieganski — APPROVED** [review]

**Comment by kamilrakoczy** [comment]
PR opened to mainline.

**Pull request closed without merging**